core/src/main/java/nu/yona/server/subscriptions/service/BuddyService.java

```diff
@@ -118,15 +127,29 @@ void clearMessagesAndSendDropBuddyMessage(User requestingUser, Buddy requestingU
 			// buddy account was removed in the meantime; nothing to do
 			return;
 		}
-		removeMessagesFromUser(requestingUserBuddy, requestingUser.getVPNLoginID());
-		sendDropBuddyMessage(requestingUser, requestingUserBuddy, message, reason);
+
+		if (requestingUserBuddy.getSendingStatus() == Status.ACCEPTED
+				|| requestingUserBuddy.getReceivingStatus() == Status.ACCEPTED)
+		{
+			removeNamedMessagesSentByUser(requestingUserBuddy.getUser(), requestingUser.getVPNLoginID());
+
+			UserAnonymized userAnonymized = UserAnonymized.getRepository().findOne(requestingUserBuddy.getVPNLoginID());
+			disconnectBuddy(userAnonymized, requestingUser.getVPNLoginID());
+			removeAnonymousMessagesSentByUser(userAnonymized, requestingUser.getVPNLoginID());
+			sendDropBuddyMessage(requestingUser, requestingUserBuddy, message, reason);
+		}
+		else
+		{
+			// remove sent buddy request message
+			removeNamedMessagesSentByUser(requestingUserBuddy.getUser(), requestingUser.getVPNLoginID());
```

Comment:
This line is identical to the first one if the if-clause, so it can go before the if-statement.